### PR TITLE
feat: update storage validation rules

### PR DESCRIPTION
**Pull request by highskore**
RHI-1293

### CODE REVIEW

**Comment by highskore** [comment]
RHI-1293

File: src/SpecsParser.sol [hunk 1]
```diff
@@ -24,12 +24,7 @@ library ERC4337SpecsParser {
 
     // Emitted if an invalid storage location is accessed
     error InvalidStorageLocation(
-        address contractAddress,
-        string contractLabel,
-        bytes32 slot,
-        bytes32 previousValue,
-        bytes32 newValue,
```
Comment (kopy-kat) [inline src/SpecsParser.sol:31]: any reason to get rid of these?

File: foundry.toml [hunk 2]
```diff
@@ -5,6 +5,7 @@ libs = ["node_modules", "lib"]
 gas_limit = "18446744073709551615"
 memory_limit = 2147483648
 verbosity = 3
+evm_version = "cancun"
```
Comment (kopy-kat) [inline foundry.toml:8]: would this be an issue for people using prev evm versions and importing this into their repo?

File: src/SpecsParser.sol [hunk 3]
```diff
@@ -238,34 +248,89 @@ library ERC4337SpecsParser {
...
+                    || (
+                        currentAccessAccount == userOpDetails.entryPoint
+                            && entities.account != address(0)
+                    );
+                // [STO-022] There is an initCode and the factory contract is staked.
+                bool isFactoryStaked = entities.isFactoryStaked;
+                if (
+                    !(
+                        isAssociatedStorage(currentSlot, currentAccessAccount, entities)
+                            && (accountAlreadyExists || isFactoryStaked)
```
Comment (kopy-kat) [inline src/SpecsParser.sol:295]: if factory is staked then only storage associated with the factory is allowed right? eg wouldnt this not revert if someone tries to access associated storage with account that does not exist but factory is staked?

File: src/SpecsParser.sol [hunk 4]
```diff
@@ -238,34 +248,89 @@ library ERC4337SpecsParser {
...
+                    continue;
+                }
+                // [STO-031] Access the entity’s own storage.
+                else if (currentAccessAccount == userOpDetails.sender) {
+                    continue;
+                }
+                // [STO-032] Read/Write Access to storage slots that are associated with the entity,
+                // in any non-entity contract.
+                else if (
+                    !isEntity(entities, currentAccessAccount)
```
Comment (kopy-kat) [inline src/SpecsParser.sol:320]: when would this be true? arent we in the `isEntityAndStaked(entities, currentAccessAccount)` condition here?

File: src/SpecsParser.sol (hunk 1 above)
Comment (highskore) [reply src/SpecsParser.sol:31]: I don't think theres a way to get the previous value using the new cheatcode (you just read opcodes), we can keep newVakue I guess

File: foundry.toml (hunk 2 above)
Comment (highskore) [reply foundry.toml:8]: It's just to support tstore checks, if they use < versions it should still be fine (only the test actually use tstore verbatim)

File: src/SpecsParser.sol (hunk 3 above)
Comment (highskore) [reply src/SpecsParser.sol:295]: true, mixed up currentAccessAccount with "account" in the specs

File: src/SpecsParser.sol (hunk 4 above)
Comment (highskore) [reply src/SpecsParser.sol]: the outer condition should actually just check if either entity is staked. (the wording is a bit werid on the ERC for this condition but I think this should be correct. fixed in [5165b53]

File: src/SpecsParser.sol (hunk 3 above)
Comment (highskore) [reply src/SpecsParser.sol]: fixed in [5165b53]. Added a new version of isAssociatedStorage that checks for association with a specific entity, the old version checked for association with any entity

File: src/SpecsParser.sol (hunk 4 above)
Comment (highskore) [reply src/SpecsParser.sol]: reworked the logic in [2978e2d]

**Review by kopy-kat — APPROVED** [review]
looks good to me

**Pull request merged**